swan-lake/development-tutorials/test-document-the-code/test-ballerina-code/execute-tests-in-parallel.md

```diff
@@ -0,0 +1,108 @@
+---
+layout: ballerina-testing-code-left-nav-pages-swanlake
+title: Execute tests in parallel
+description: Learn how to execute tests in parallel.
+keywords: ballerina, programming language, testing, test execution
+permalink: /learn/test-ballerina-code/execute-tests-in-parallel/
+active: execute-tests-in-parallel
+intro: Ballerina test framework allows developers to execute test cases in parallel, which can significantly reduce the overall test execution time, especially for large codebases with a large number of test cases.
+---
+
+## Enable parallel execution of tests
+
+Tests are executed serially by default. The user has to explicitly activate the parallel execution by providing the `--parallel` flag as follows.
+
+```
+bal test --parallel
+```
+
+Before running tests concurrently, it is crucial to ensure concurrency safety. Ballerina inherently evaluates the concurrency safety of tests to some extent using a predefined set of rules. Tests that do not comply with these rules are executed sequentially, regardless of the parallel flag being enabled.
+
+## Exclude a specific test from parallel execution
+
+Certain tests may have dependencies or requirements that make them incompatible with parallel execution, such as shared resources or specific environment setups. If such a requirement is identified, the user may need to set the `serialExecution` flag to `true` as follows:
+
+```ballerina
+@test:Config {serialExecution: true}
+function testAssertEquals6() {
+    test:assertEquals(100, 100);
+}
+```
+
+## Write a concurrent safe test case
+
+A set of rules must be followed while writing a parallel test.
+
+1. The test function must be isolated. In some instances, the compiler infers the test functions as isolated automatically if there are sufficient conditions to make them concurrently safe.
+2. If it is a data provider test, 
+   - The data provider of the test function must be isolated.
+   - The test function parameters must be read-only type.
```

Comment:
The types of the parameters have to be subtypes of `readonly`. I.e., only immutable values can be passed as arguments.